java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -23,219 +23,260 @@
-    MethodInvocationTree mit = (MethodInvocationTree) invocationNode.programPoint.syntaxTree();
-    List<SymbolicValue> arguments = argumentsUsedForMethodInvocation(invocationNode, mit);
-    return arguments.indexOf(candidate);
-  }
+    private Stream<JavaFileScannerContext.Location> methodInvocationFlow(Constraint learnedConstraint, ExplodedGraph.Edge edge) {
+      ExplodedGraph.Node parent = edge.parent;
+      MethodInvocationTree mit = (MethodInvocationTree) parent.programPoint.syntaxTree();
+      Stream.Builder<JavaFileScannerContext.Location> flowBuilder = Stream.builder();
+      SymbolicValue returnSV = edge.child.programState.peekValue();
+      if (symbolicValues.contains(returnSV)) {
+        flowBuilder.add(location(parent, String.format("'%s()' returns %s.", mit.symbol().name(), learnedConstraint.valueAsString())));
+      }
+      SymbolicValue invocationTarget = parent.programState.peekValue(mit.arguments().size());
+      if (symbolicValues.contains(invocationTarget)) {
+        flowBuilder.add(location(parent, "..."));
+      }
 
-  private static List<SymbolicValue> argumentsUsedForMethodInvocation(ExplodedGraph.Node invocationNode, MethodInvocationTree mit) {
-    List<SymbolicValue> values = invocationNode.programState.peekValues(mit.arguments().size());
-    return Lists.reverse(values);
-  }
-  @Nullable
-  private Symbol addFlowFromLearnedSymbols(ExplodedGraph.Node currentNode, @Nullable Symbol trackSymbol) {
-    ExplodedGraph.Node parent = currentNode.parent();
-    if (trackSymbol == null || parent == null) {
-      return null;
-    }
-    Optional<LearnedAssociation> learnedAssociation = currentNode.learnedAssociations()
-      .filter(lv -> lv.symbol().equals(trackSymbol))
-      .findFirst();
-    if (learnedAssociation.isPresent()) {
-      LearnedAssociation la = learnedAssociation.get();
-      for (Class<? extends Constraint> domain : domains) {
-        Constraint constraint = parent.programState.getConstraint(la.symbolicValue(), domain);
-        if(constraint != null) {
-          flow.add(location(parent, String.format("'%s' is assigned %s.", la.symbol().name(), constraint.valueAsString())));
+      List<Integer> argumentIndices = correspondingArgumentIndices(symbolicValues, parent);
+      argumentIndices.stream()
+        .map(mit.arguments()::get)
+        .map(argTree -> {
+          String message = String.format("Implies '%s' is %s.", SyntaxTreeNameFinder.getName(argTree), learnedConstraint.valueAsString());
+          return new JavaFileScannerContext.Location(message, argTree);
+        })
+        .forEach(flowBuilder::add);
+
+      MethodYield selectedMethodYield = edge.child.selectedMethodYield(parent);
+      if (selectedMethodYield != null) {
+        if (symbolicValues.contains(returnSV)) {
+          // to retrieve flow for return value
+          argumentIndices.add(-1);
         }
+        // FIXME SONARJAVA-2076 consider all flows from yields
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)